File: frontend/src/services/productService.ts
// services/productService.ts
import apiClient from '../lib/axios';
import { PageResponse } from '../types/PageResponse';
import { Product, ProductDetailResponse, ProductQueryParams, ProductAutoComplete, ProductVariantAutoComplete, BOMItemResponse } from '../types/product';
import qs from 'qs'; 


// to fetch the product after searching + filtering
export const getProducts = async (
  params: ProductQueryParams
  
): Promise<PageResponse<Product>> => {
  const defaultParams = {
    page: 0,      
    size: 10,     
    search: '',   
    groupProduct: [], 
    status: []        
  };
  const mergedParams = { ...defaultParams, ...params };

  console.log('Sending params to /products/filter', mergedParams);
  const response = await apiClient.get('/products/filter', { 
    params: mergedParams,
    paramsSerializer: (params) =>
      qs.stringify(params, { arrayFormat: 'repeat' }) 
      // result: ?groupProduct=A&groupProduct=B&status=ACTIVE
  });
  return response.data;
};

// to fetch the product for auto complete while searching
export const getAllProductsForAutocomplete = async (): Promise<ProductAutoComplete[]> => {
  const response = await apiClient.get('/products/all');
  return response.data;
};

export const getAllProductVariantsForAutocomplete = async (search?: string): Promise<ProductVariantAutoComplete[]> => {
  const params = search ? `?search=${encodeURIComponent(search)}` : '';
  const response = await apiClient.get(`/products/variants/autocomplete${params}`);
  return response.data;
};

// to fetch the detail product page 
export const getProductById = async (id: number): Promise<ProductDetailResponse> => {
  const res = await apiClient.get<ProductDetailResponse>(`/products/${id}`);
  return res.data;
};

// to update product
export const updateProduct = async (id: number, data: any): Promise<any> => {
  try {
    const response = await apiClient.put(`/products/${id}`, data);
    return response.data;
  } catch (error) {
    throw error;
  }
};

// to get all product groups
export const getAllProductGroups = async (): Promise<any[]> => {
  try {
    const response = await apiClient.get('/products/product-groups');
    return response.data;
  } catch (error) {
    throw error;
  }
};

export const getAvailableStatusTransitions = async (productId: number): Promise<any> => {
  try {
    const response = await apiClient.get(`/products/${productId}/status-transitions`);
    return response.data;
  } catch (error) {
    throw error;
  }
};

export const changeProductStatus = async (productId: number, newStatus: string, reason?: string): Promise<any> => {
  try {
    const response = await apiClient.put(`/products/${productId}/status`, {
      newStatus,
      reason
    });
    return response.data;
  } catch (error) {
    throw error;
  }
};

// to get BOM items for a specific product variant
export const getBomItem = async (productId: number, variantId: number): Promise<BOMItemResponse[]> => {
  try {
    const response = await apiClient.get<BOMItemResponse[]>(`/products/${productId}/variant/${variantId}`);
    return response.data;
  } catch (error) {
    throw error;
  }
};

// to delete a product (soft delete)
export const deleteProduct = async (id: number): Promise<any> => {
  try {
    const response = await apiClient.delete(`/products/${id}`);
    return response.data;
  } catch (error) {
    throw error;
  }
};

File: backend/src/main/java/org/example/coretrack/service/ProductService.java
package org.example.coretrack.service;

import java.util.List;

import org.example.coretrack.dto.product.AddProductRequest;
import org.example.coretrack.dto.product.AddProductResponse;
import org.example.coretrack.dto.product.AllProductSearchResponse;
import org.example.coretrack.dto.product.BOMItemResponse;
import org.example.coretrack.dto.product.ChangeProductStatusRequest;
import org.example.coretrack.dto.product.ChangeProductStatusResponse;
import org.example.coretrack.dto.product.ProductDetailResponse;
import org.example.coretrack.dto.product.ProductGroupResponse;
import org.example.coretrack.dto.product.ProductStatusTransitionResponse;
import org.example.coretrack.dto.product.ProductVariantAutoCompleteResponse;
import org.example.coretrack.dto.product.SearchProductResponse;
import org.example.coretrack.dto.product.UpdateProductRequest;
import org.example.coretrack.dto.product.UpdateProductResponse;
import org.example.coretrack.dto.product.DeleteProductResponse;
import org.example.coretrack.model.auth.User;
import org.example.coretrack.model.auth.Company;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {
    // to add a new product to product catalog
    AddProductResponse createProduct(AddProductRequest addProductRequest, User createdByUser);
    
    // to filter the product that meet the criteria (with company context)
    Page<SearchProductResponse> findProduct(
            String search,
            List<String> groupProducts,
            List<String> statuses,
            Pageable pageable,
            User currentUser);

    // to return all products (with company context)
    Page<SearchProductResponse> findAllProducts(Pageable pageable, User currentUser);

    // to return all product while searching (with company context)
    List<AllProductSearchResponse> getAllProductsForAutocomplete(String search, User currentUser);

    // to find product by ID (with company context)
    ProductDetailResponse getProductById(Long id, User currentUser);

    // to get all product groups (with company context)
    List<ProductGroupResponse> getAllGroupName(User currentUser);

    // to update a product (with company context)
    UpdateProductResponse updateProduct(Long id, UpdateProductRequest request, User updatedByUser);
    
    // to change product status (with company context)
    ChangeProductStatusResponse changeProductStatus(Long productId, ChangeProductStatusRequest request, User changedByUser);
    
    // to get available status transitions (with company context)
    ProductStatusTransitionResponse getAvailableStatusTransitions(Long productId, User currentUser);
    
    // to return all product variants for autocomplete (with company context)
    List<ProductVariantAutoCompleteResponse> getAllProductVariantsForAutocomplete(String search, User currentUser);

    // to return BomItem (with company context)
    List<BOMItemResponse> getBomItem(Long id, Long variantId, User currentUser);
    
    // to delete a product (soft delete) (with company context)
    DeleteProductResponse deleteProduct(Long id, User deletedByUser);
}
